### PR TITLE
Remove AS-NS linking in the AS and NS

### LINKED ISSUE #2312: Reverse AS-NS linking direction
#### Background
_Can be skipped if you are familiar with https://github.com/TheThingsIndustries/lorawan-stack/issues/941_

In the current implementation of the communication between the Network Server and Application Server, the links are established by the Application Server, which links itself to the Network Server. Over this link, uplinks are sent as a stream of `ttnpb.ApplicationUp`.

The underlying issue with this approach is that Application Servers are inherently stateful due to linking. This introduces challenges with regards to load balancing between multiple Application Servers - the distribution of links over multiple instances, maintaining/migrating links when instances are down or a network partition occurs being just a number of issues introduced by the linking state.

#### Proposed solution

In order to tackle this issue, following our offline discussions, we've decided that we would like to reverse the direction of the links. This means that instead of the Application Server being a client to the Network Server, the Network Server becomes a client of the Application Server. This simplifies the logic for both components.

Since the Application Server may have clients (MQTT/gRPC) connect to it while not being linked by a Network Server, a new PubSub service should be used as a broker between Application Server instances. When a new application subscription (connection) arrives to an Application Server, a PubSub subscription is created for the uplinks of the said application. When an uplink is received by an Application Server instance, it is published to the topic of the application.

#### Implementation details:

- `HandleUplink` should be added to a new `NsAs` service that the Application Server implements.
```proto
service NsAs {
  rpc HandleUplink(ApplicationUp) returns (google.protobuf.Empty);
}
```
> We may want to revisit https://github.com/TheThingsNetwork/lorawan-stack/issues/1523 since we are already adding a new RPC.
- The NS uses this RPC in order to push uplinks to the AS.
- The NS uses `component.GetPeer`/`component.GetPeerCon` in order to receive the peer/connection to the AS ~based on the application identifiers~.
- When the AS receives an uplink, it triggers the required webhooks and application packages. Only the AS that receives the uplink should send the webhooks and run the application packages. 
The uplink is then published to the PubSub service, from where subscription based frontends can pick it up.
- AS downlink queue operations no longer use the link connection, but rather the cluster peer connection.
- AS can use `gocloud.dev/pubsub` for PubSub communication for inter-AS PubSub.

#### Required migrations
We need to migrate the application default payload formatters to a more general Application Server application store. The migration should be lazy - when we need the application payload formatter we check if we have it in the AS application store, and if it is missing we attempt to retrieve it from the old link store. Any writes/updates are done in the application store.

#### Open questions
- ~Should we want to drop external NS linking completely ?~
- Which PubSub service should we bundle by default ? I personally propose NATS, but I do not have strong feelings about this.
- ~How can AS PubSub integrations handle downlink operations ? Internally we can ensure that uplinks are sent only once, using subscription groups, but I don't see how we can deduplicate downlink queue operations, since some services (MQTT) do not support subscription groups.~
Reworded: How can we decide which instance starts an external PubSub integration (NATS/MQTT), given that having each instance connect to every integration would result in duplicated downlink queue operations ?

### LINKED ISSUE #3145: Flaky test applicationserver linking_test
#### Summary

https://github.com/TheThingsNetwork/lorawan-stack/pull/3139/checks?check_run_id=1026896468

### PR DESCRIPTION
**Pull request by adriansmares**
#### Summary

Closes #2312 
References #3190 
References #3260 
Closes https://github.com/TheThingsNetwork/lorawan-stack/issues/3145

#### Changes

- Reintroduce the changes added by #3190

#### Testing

Covered by #3190 

##### Regressions

Covered by #3190 

#### Notes for Reviewers

I merged #3190 too fast, which left the `v3.10` branch in an unusable state. As such, I've reverted the merge in #3260 in order to make the branch usable again. 

This PR was generated using `git revert -m 1 62f404a75c37a93b431b0133927bcfd773404c4f` on 62f404a75.

As it was already approved, this PR should be used as a base for the NS changes, once @rvolosatovs has enough time to pick them up. We can then merge them into this branch and then into `v3.10`.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Comment by johanstokking** [comment]
OK. So what is the plan here, @rvolosatovs takes this over?

**Comment by adriansmares** [comment]
> 
> 
> OK. So what is the plan here, @rvolosatovs takes this over?

Yes.

File: api/applicationserver.proto
Comment (htdvisser) [inline api/applicationserver.proto:117]: We should probably describe this behavior change in CHANGELOG

File: cmd/internal/shared/applicationserver/config.go
Comment (htdvisser) [inline cmd/internal/shared/applicationserver/config.go:34]: We should probably describe this behavior change in CHANGELOG

File: pkg/applicationserver/config.go
Comment (htdvisser) [inline pkg/applicationserver/config.go]: Please document configuration changes

Comment (htdvisser) [reply pkg/applicationserver/config.go]: Also, we have to add a dummy flag `--as.link-mode` to avoid breaking existing deployments that do use that flag.

**Comment by rvolosatovs** [comment]
@adriansmares please address the comments, since that's something missed in #3190

**Comment by adriansmares** [comment]
@rvolosatovs I've rebased this on top of 0b16aaf6a02b0980e9b14d7ba0911ed58593bfcf and addressed @htdvisser comments.

**Comment by rvolosatovs** [comment]
> @rvolosatovs I've rebased this on top of [0b16aaf](https://github.com/TheThingsNetwork/lorawan-stack/commit/0b16aaf6a02b0980e9b14d7ba0911ed58593bfcf) and addressed @htdvisser comments.

Awesome, thanks!

**Comment by johanstokking** [comment]
I guess this one will go to 3.11 then?

**Comment by htdvisser** [comment]
Definitely not v3.10. As far as I know this isn't finished on the NS side yet.

**Comment by rvolosatovs** [comment]
It's in-progress, why not 3.10?

**Comment by htdvisser** [comment]
1. The deadline for non-compatible v3.10 changes was 5 minutes ago
2. It's a bit risky, because of the many other changes in the NS

**Review by johanstokking — COMMENTED** [review]
Please request review when ready @rvolosatovs

**Comment by rvolosatovs** [comment]
> Please request review when ready @rvolosatovs

I will, but not on this PR; on https://github.com/TheThingsNetwork/lorawan-stack/pull/3398

**Comment by rvolosatovs** [comment]
blocked by #3398

**Comment by johanstokking** [comment]
I'm assuming the review will be #3398 when #3445 is reviewed and merged. Please confirm @adriansmares @rvolosatovs

**Comment by adriansmares** [comment]
> I'm assuming the review will be #3398 when #3445 is reviewed and merged. Please confirm @adriansmares @rvolosatovs

This is correct.

**Comment by rvolosatovs** [comment]
This is not blocked anymore, right?

**Comment by rvolosatovs** [comment]
@kschiffer can you give some insight in why end-to-end tests are failing?

**Comment by kschiffer** [comment]
See here #3464

**Comment by johanstokking** [comment]
@rvolosatovs what should be reviewed here? I feel like the original changes from @adriansmares are reviewed and the other NS changes coming in here too.

**Comment by rvolosatovs** [comment]
The NS changes are already reviewed
The AS changes I believe as well
I think we can merge this as long as it's tested?

**Comment by rvolosatovs** [comment]
@adriansmares please rebase

**Comment by adriansmares** [comment]
@rvolosatovs I've rebased this and it works end-to-end ! with a small diff in the NS that you can find bellow:
- Change authentication for queue list/replace/push to `clusterauth` - I hope this is obvious.
- Store the `RawPayload` of recent uplinks - when I first started testing I had lots of:

```
ERROR Failed to determine transmission number, skip ack=false adr=true adr_ack_req=false bandwidth=125000 class_b=false dev_addr=008D48FD device_uid=app1.dev1p error=error:pkg/networkserver:raw_payload_too_short (length of RawPayload must not be less than 4) f_cnt_reset=false f_opts_len=0 f_port=102 frequency=868100000 full_f_cnt_up=35 grpc_method=HandleUplink grpc_service=ttn.lorawan.v3.GsNs last_f_cnt_up=0 m_type=UNCONFIRMED_UP mac_version=1.0.3 major=LORAWAN_R1 max_transmissions=1 namespace=networkserver pending_session=false phy_payload_len=14 received_at=2020-11-20 18:46:10.537956896 +0000 UTC request_id=01EQKFMRV95FX9JTT7PQPSMTG5 spreading_factor=7 uplink_f_cnt=35
```

And I found out that basically `transmissionNumber` requires the raw payload of the recent uplinks, but it is not saved anymore (not sure which PR removed this). You have more context than me on this one, so feel free to fix it your own way.

```diff
diff --git a/pkg/networkserver/grpc_asns.go b/pkg/networkserver/grpc_asns.go
index d83cc79c4..c0d4537fa 100644
--- a/pkg/networkserver/grpc_asns.go
+++ b/pkg/networkserver/grpc_asns.go
@@ -21,7 +21,7 @@ import (
        "time"
 
        pbtypes "github.com/gogo/protobuf/types"
-       "go.thethings.network/lorawan-stack/v3/pkg/auth/rights"
+       clusterauth "go.thethings.network/lorawan-stack/v3/pkg/auth/cluster"
        "go.thethings.network/lorawan-stack/v3/pkg/errors"
        "go.thethings.network/lorawan-stack/v3/pkg/events"
        "go.thethings.network/lorawan-stack/v3/pkg/frequencyplans"
@@ -231,7 +231,7 @@ var errDownlinkQueueCapacityExceeded = errors.DefineResourceExhausted("downlink_
 func (ns *NetworkServer) DownlinkQueueReplace(ctx context.Context, req *ttnpb.DownlinkQueueRequest) (*pbtypes.Empty, error) {
        if n := len(req.Downlinks); n > ns.downlinkQueueCapacity*2 {
                return nil, errDownlinkQueueCapacityExceeded.New()
-       } else if err := rights.RequireApplication(ctx, req.ApplicationIdentifiers, ttnpb.RIGHT_APPLICATION_LINK); err != nil {
+       } else if err := clusterauth.Authorized(ctx); err != nil {
                return nil, err
        }
 
@@ -302,7 +302,7 @@ func (ns *NetworkServer) DownlinkQueuePush(ctx context.Context, req *ttnpb.Downl
                return nil, errDownlinkQueueCapacityExceeded.New()
        } else if n == 0 {
                return ttnpb.Empty, nil
-       } else if err := rights.RequireApplication(ctx, req.ApplicationIdentifiers, ttnpb.RIGHT_APPLICATION_LINK); err != nil {
+       } else if err := clusterauth.Authorized(ctx); err != nil {
                return nil, err
        }
 
@@ -356,7 +356,7 @@ func (ns *NetworkServer) DownlinkQueuePush(ctx context.Context, req *ttnpb.Downl
 
 // DownlinkQueueList is called by the Application Server to get the current state of the downlink queue for a device.
 func (ns *NetworkServer) DownlinkQueueList(ctx context.Context, ids *ttnpb.EndDeviceIdentifiers) (*ttnpb.ApplicationDownlinks, error) {
-       if err := rights.RequireApplication(ctx, ids.ApplicationIdentifiers, ttnpb.RIGHT_APPLICATION_LINK); err != nil {
+       if err := clusterauth.Authorized(ctx); err != nil {
                return nil, err
        }
        dev, ctx, err := ns.devices.GetByID(ctx, ids.ApplicationIdentifiers, ids.DeviceID, []string{
diff --git a/pkg/networkserver/grpc_gsns.go b/pkg/networkserver/grpc_gsns.go
index 93f3c3104..4a6966b1c 100644
--- a/pkg/networkserver/grpc_gsns.go
+++ b/pkg/networkserver/grpc_gsns.go
@@ -970,6 +970,7 @@ func (ns *NetworkServer) handleDataUplink(ctx context.Context, up *ttnpb.UplinkM
                                "mac_state.recent_uplinks",
                        )
                        stored.MACState.RecentUplinks = appendRecentUplink(stored.MACState.RecentUplinks, &ttnpb.UplinkMessage{
+                               RawPayload:         up.RawPayload,
                                Payload:            up.Payload,
                                Settings:           up.Settings,
                                RxMetadata:         up.RxMetadata,
```

**Comment by rvolosatovs** [comment]
@adriansmares authorization diff LGTM and I will fix the payload stuff in a separate PR

**Comment by adriansmares** [comment]
@rvolosatovs I've rebased this on the latest `v3.11` and 'fixed' the tests. Please take a look and if everything is fine there this PR is basically ready to merge (still needs an approval but everything has been basically reviewed at this point).

**Review by rvolosatovs — APPROVED** [review]

**Comment by rvolosatovs** [comment]
@johanstokking

File: pkg/networkserver/grpc_asns_test.go
Comment (rvolosatovs) [inline pkg/networkserver/grpc_asns_test.go:57]: Can you remove the `ContextFunc` fields in these structs as well?

**Comment by rvolosatovs** [comment]
blocking https://github.com/TheThingsIndustries/lorawan-stack/pull/2419

**Review by johanstokking — APPROVED** [review]

**Review by htdvisser — APPROVED** [review]
Code-owned files LGTM

**Pull request merged**
